Add Rekor functionality

jyotsna-penumaka opened the pull request.

jyotsna-penumaka committed:
Use openapi-generator-cli to convert the openapi.yml file to .rs code.

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Create README.md

jyotsna-penumaka committed:
Update Readme

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add structures

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add gitignore files

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Remove node_modules

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Remove node_modules

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Removed node_modules

jyotsna-penumaka committed:
Define structures to read the log entry after creating one

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Removed target folder from test-rekor folder

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add return structs and tests for create_log_entry, get_log_entry_by_index, get_log_entry_by_uuid

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add return structs and tests for create_log_entry, get_log_entry_by_index, get_log_entry_by_uuid

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add tests and return structs for search_log_query, search_index, get_public_key, get_rekor_version, get_timestamp_cert_chain, get_timestamp_response, get_log_info, get_log_proof

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Clean up

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add test for get_timestamp_response

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Trying to test get_timestamp_response

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Debug Test#9 : get_timestamp_response

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Clean up + Specify the code that needs review

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Test the Rekor API and add structures to hold the responses

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Merge branch 'master' into rekor-functionality

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Code clean up

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Fix bug in tlog_api

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add file permissions.
Resolves #1

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

lkatalin commented on rekor-rs/.gitignore:
```diff
@@ -1,3 +1,14 @@
+
+# Added by cargo
+
+/target
```
Is this one different from `/target/`?

jyotsna-penumaka replied:
You are right, I was referring to the same directory.

lkatalin commented on rekor-rs/src/apis/entries_api.rs:
```diff
@@ -72,17 +78,22 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
     let local_var_status = local_var_resp.status();
     let local_var_content = local_var_resp.text().await?;
-
+    
```
Whitespace ^

jyotsna-penumaka replied:
I am not very clear on what to do here.
Do I have an extra whitespace somewhere?

lkatalin replied:
Yeah, if you see the green line (your change) you've got that darker green rectangle on the left. That indicates a set of spaces or a tab that you inserted on that line. If you do a `git diff upstream/main myfile.rs` you'll be able to see this change locally and you can go and delete it.

lkatalin commented on rekor-rs/src/apis/entries_api.rs:
```diff
@@ -72,17 +78,22 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Format the returned response such that it can be read into a struct
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
```
I know I'm the one who suggested parsing it this way, but there is probably a better way to do it that is more readable. Once the `create_log_entry` functionality is working properly it would be good to try it a different way, maybe splitting the string on certain key words or something and/or making the parsing into its own function.

jyotsna-penumaka replied:
Once the following issue gets resolved, we can get rid of the parsing all together.
sigstore/rekor#808
(Thanks @lkatalin for letting me know about this issue)

lkatalin commented on rekor-rs/src/apis/entries_api.rs:
```diff
@@ -102,7 +113,11 @@ pub async fn get_log_entry_by_index(configuration: &configuration::Configuration
+        // Format the returned response such that it can be read into a struct
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+        let sum = "{\"uuid\": ".to_string() + &(uuid.to_owned()) + "," + rest;
```
Same as above

lkatalin commented on rekor-rs/src/apis/entries_api.rs:
```diff
@@ -130,15 +145,21 @@ pub async fn get_log_entry_by_uuid(configuration: &configuration::Configuration,
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Format the returned response such that it can be read into a struct
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
```
Same as above

lkatalin commented on rekor-rs/src/apis/timestamp_api.rs:
```diff
@@ -73,21 +79,33 @@ pub async fn get_timestamp_response(configuration: &configuration::Configuration
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Save the response returned by rekor in a file and return the path
+        let file_name = "response.tsr";
```
It might be good to make the file name a constant.

lkatalin commented on rekor-rs/src/models/log_entry.rs:
```diff
@@ -0,0 +1,69 @@
+    attestation: Option<Attestation>,
+    body: String,
+    integrated_time: i64,
+    log_i_d: String,
```
Maybe call it `log_id`?

jyotsna-penumaka replied:
The response we get from rekor looks like this 	
`"logID": "c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d",`
Since rust uses snake case I named the field ad `log_i_d`.
We serialize the fields as camelCase before we use serde_json to read the response into the struct.

Changing the field from `log_i_d` to `log_id` causes the following error:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Serde(Error("missing field `logId`", line: 1, column: 7880))', examples/get_log_entry_by_index.rs:9:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

lkatalin replied:
Ah, gotcha. In that case, what you have is better.

lkatalin commented on rekor-rs/src/models/log_entry.rs:
```diff
@@ -0,0 +1,69 @@
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct LogEntry {
+    // native uuid object maybe?
```
Yeah, not crucial atm but a `uuid` type could be good so that we can have methods on it that validate it properly.

jyotsna-penumaka replied:
If I create a separate struct for uuid, it would only have a single string field. 
Do you think its still fine?

lkatalin replied:
I'd say it is fine, in this case the struct is less a "collection of data" and more like a wrapper so that you can define specific methods on it like `is_valid()`. However, I'm not sure what I commented on here because I don't see `uuid` hanging out as a bunch of numbers...

lkatalin replied:
Oh wait, I see it as a String. : )

lkatalin commented on rekor-rs/src/models/log_entry.rs:
```diff
@@ -0,0 +1,69 @@
+    verification: Verification,
+}
+
+
```
Whitespace

lkatalin commented on rekor-rs/src/models/rekord.rs:
```diff
@@ -33,4 +34,80 @@ impl Rekord {
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PublicKey {
+    content: String,
```
It could be good to make this an actual key type at some point (using the `openssl` crate or similar) because it would help validate it, but this seems fine for now.

jyotsna-penumaka replied:
I think the contents field is supposed to hold in the base64 data inside a pem key file.
`Pem` crate looks promising https://docs.rs/pem/latest/pem/struct.Pem.html 
The only problem is that the header and the footer for the pem file is missing.

lkatalin replied:
Not sure what you mean that it's missing, but let me know if you want to talk through it more.

lkatalin commented on rekor-rs/src/models/rekord.rs:
```diff
@@ -33,4 +34,80 @@ impl Rekord {
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Hash {
+    algorithm: String,
```
The algo could also be an enum of valid algorithms eventually.

jyotsna-penumaka replied:
https://github.com/sigstore/rekor/blob/main/cmd/rekor-cli/app/pflags.go#L169 
https://github.com/sigstore/rekor/blob/main/cmd/rekor-cli/app/search.go#L58 
suggest that sha256 and sha1 are the 2 valid formats.
I changed the type of algorithm from String to an enum, with sha256 and sha1 being the two variants of the enum.

lkatalin replied:
Perfect!

lkatalin commented on README.md:
````diff
@@ -0,0 +1,6 @@
+```
+cd test_rekor
+cargo run
+```
````
If there's a way these can be invoked with `cargo test`, that is preferable. That way we can call each test individually for each function.

jyotsna-penumaka replied:
I thought an examples folder would be better for this repo.
In order to view the available examples:
`$ cargo run --example`
To run individual examples and view the output you can do:
`$ cargo run --example get_log_entry_by_index`

lkatalin reviewed (commented):
This looks pretty good to me, nothing major aside from suggesting a structure where the integration tests can be run individually with `cargo test testname`. I can see you still have some open todos, but this seems good re: the code you already have. Also I tried the tests and they did work on my end.

jyotsna-penumaka committed:
Add examples to Rekor-rs library

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update README.md with instructions to run the examples

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Change algorithm’s type from String to an enum that takes sha256 or sha1

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Modify .gitignore

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Allow non_camel_case_types for AlgorithmKind Enum

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add Apache License

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add github actions

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update github actions

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update git hub actions

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Fix the bugs in the actions workflow

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update github actions workflow

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Run cargo fmt and fix the github actions

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Fix github actions for cargo fmt

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Restructure the repo

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Restructure the repo and debug get_log_info and get_log_proof

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Restructure the repo

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Restructure the data

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update the manifest file paths in github actions

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update readme

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Remove target files

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update .gitignore

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Make clippy changes

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Remove the unnecessary use of `to_owned` in src/apis/entries_api.rs

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Update README.md

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Remove reference which is immediately dereferenced

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Restructure rekor-rs repo

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Merge branch 'master' into rekor-functionality

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,56 @@
+    let configuration = Configuration::default();
+    let hash = Hash::new(
+        AlgorithmKind::sha256,
+        "e2535d638859bb63ea9ea5cf467562cba63b007eae1acd0d73a3f259c582561f".to_string(),
```
It would be helpful to have a short comment for each of these data blobs so it is easy to understand their role in the examples. It would also be helpful to have an info comment at the top of each example file explaining what it does (it could be copied partly from the comments you have on the corresponding functions).

flavio replied:
I would even expect the user to provide this data and have the example code deal with producing all these blobs. After all, the purpose of the examples is IMHO to provide step-by-step instructions that other developers can almost copy and paste.

If this is not possible, adding some explanations like @lkatalin suggested would be my recommendation

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,34 @@
+        url: None
+    };
+    let query = SearchIndex {
+        email: Some("jpenumak@redhat.com".to_string()),
```
Do you want your email to become part of public code? : ) We can figure out some way around this.

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -52,18 +51,29 @@ pub enum SearchLogQueryError {
-/// Creates an entry in the transparency log for a detached signature, public key, and content. Items can be included in the request or fetched by the server when URLs are specified. 
-pub async fn create_log_entry(configuration: &configuration::Configuration, proposed_entry: crate::models::ProposedEntry) -> Result<::std::collections::HashMap<String, serde_json::Value>, Error<CreateLogEntryError>> {
+/// Creates an entry in the transparency log for a detached signature, public key, and content. Items can be included in the request or fetched by the server when URLs are specified.
+// Change the return value of the function to LogEntry from ::std::collections::HashMap<String, serde_json::Value>
```
Is this a TODO? Looks like it's done!

jyotsna-penumaka replied:
It was a TODO earlier.
Thanks for pointing it out :100:

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -74,25 +84,40 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
 
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Format the returned response such that it can be read into a struct
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+        let sum = "{\"uuid\": ".to_string() + uuid + "," + rest;
+        serde_json::from_str::<LogEntry>(&sum).map_err(Error::from)
```
I guess this can be cleaned up once the Rekor issue around the dynamic top-level key is closed. Probably needs a reminder issue or at least a TODO comment.

lkatalin replied:
Could you make this parsing into its own function? I see you're doing it in several places, and if it was one function it would be easier to update later once we don't have the dynamic top-level key.

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -74,25 +84,40 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
 }
 
-pub async fn get_log_entry_by_index(configuration: &configuration::Configuration, log_index: i32) -> Result<::std::collections::HashMap<String, serde_json::Value>, Error<GetLogEntryByIndexError>> {
+// Change the return value of the function to LogEntry from ::std::collections::HashMap<String, serde_json::Value>
```
Seems like this is done as well

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -102,25 +127,43 @@ pub async fn get_log_entry_by_index(configuration: &configuration::Configuration
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+        let sum = "{\"uuid\": ".to_string() + uuid + "," + rest;
+        serde_json::from_str::<LogEntry>(&sum).map_err(Error::from)
```
Same as above

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -130,24 +173,42 @@ pub async fn get_log_entry_by_uuid(configuration: &configuration::Configuration,
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+        let sum = "{\"uuid\": ".to_string() + uuid + "," + rest;
+        serde_json::from_str::<LogEntry>(&sum).map_err(Error::from)
```
Same as above

lkatalin commented on src/apis/index_api.rs:
```diff
@@ -24,17 +22,24 @@ pub enum SearchIndexError {
+        local_var_configuration.base_path
+    );
+    let mut local_var_req_builder =
+        local_var_client.request(reqwest::Method::POST, local_var_uri_str.as_str());
```
Interesting that this is POST, that's different from `get_entry_by_uuid` - are we sending data? No need to change anything, just curious

lkatalin commented on src/apis/timestamp_api.rs:
```diff
@@ -4,16 +4,22 @@
 
-use crate::apis::ResponseContent;
-use super::{Error, configuration};
+// Case 1:
```
What is case 1? Is this a leftover comment?

jyotsna-penumaka replied:
Its a leftover comment :sweat_smile:

lkatalin commented on src/apis/timestamp_api.rs:
```diff
@@ -34,60 +40,95 @@ pub enum GetTimestampResponseError {
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Directly return the certificate chain instead of using serde to convert it to a string
+        // serde_json::from_str(&local_var_content).map_err(Error::from)
```
Leftover comment

lkatalin replied:
Remove please

lkatalin commented on src/apis/timestamp_api.rs:
```diff
@@ -34,60 +40,95 @@ pub enum GetTimestampResponseError {
-        local_var_req_builder = local_var_req_builder.header(reqwest::header::USER_AGENT, local_var_user_agent.clone());
+        local_var_req_builder =
+            local_var_req_builder.header(reqwest::header::USER_AGENT, local_var_user_agent.clone());
+        // specify the header for the request
```
todo or leftover?

lkatalin replied:
Why is this commented if the others aren't?

lkatalin commented on src/apis/tlog_api.rs:
```diff
@@ -53,32 +54,47 @@ pub async fn get_log_info(configuration: &configuration::Configuration, ) -> Res
     let local_var_content = local_var_resp.text().await?;
 
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
+        println!("{:#?}", local_var_content);
```
Should this be some kind of debug info or error handled?

jyotsna-penumaka replied:
I was using it to understand how to parse the response.
I got rid of it now :smile:

lkatalin commented on src/lib.rs:
```diff
@@ -1,10 +1,10 @@
 extern crate serde;
 extern crate serde_json;
 extern crate url;
-extern crate reqwest;
```
I don't think `extern crate` is required in Rust anymore, you can just put deps in `Cargo.toml` and `use` them.

lkatalin replied:
It's still here!

lkatalin commented on src/models/inactive_shard_log_info.rs:
```diff
@@ -28,7 +25,12 @@ pub struct InactiveShardLogInfo {
-    pub fn new(root_hash: String, tree_size: i32, signed_tree_head: String, tree_id: String) -> InactiveShardLogInfo {
+    pub fn new(
+        root_hash: String,
+        tree_size: i32,
```
I think the tree size is meant to be `i64`, does that mess things up with serde? Refs:
https://github.com/sigstore/rekor/blob/547eb3cef9690bc97e73e35eb3bf1f40b18a3504/cmd/rekor-cli/app/log_info.go#L46
https://github.com/sigstore/rekor/blob/547eb3cef9690bc97e73e35eb3bf1f40b18a3504/pkg/sharding/ranges.go#L42

lkatalin replied:
Bump

lkatalin commented on src/models/log_entry.rs:
```diff
@@ -0,0 +1,68 @@
+// Design a struct that stores the response returned by Rekor after making a new entry
+
+/// Stores the response returned by Rekor after making a new entry
+// get rid of default
```
Leftover comment?

lkatalin commented on src/models/log_entry.rs:
```diff
@@ -0,0 +1,68 @@
+#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct LogEntry {
+    // native uuid object maybe?
```
I like to put the word `TODO` or similar in these types of comments because they are easy to find later with ctrl + f

lkatalin commented on src/models/log_entry.rs:
```diff
@@ -0,0 +1,68 @@
+#[serde(rename_all = "camelCase")]
+pub struct Attestation {
+    // This field is just a place holder
+    // Not sure what is stored inside the Attestation struct, it is empty for now
```
todo

lkatalin commented on src/models/log_entry.rs:
```diff
@@ -0,0 +1,68 @@
+    tree_size: i64,
+}
+
+/*
```
I would remove this example response or put it somewhere in its own file not as a comment

lkatalin commented on src/models/rekord.rs:
```diff
@@ -33,4 +32,83 @@ impl Rekord {
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Signature {
+    format: String,
```
What is the format? Would this make sense as an enum?

lkatalin commented on src/models/search_index.rs:
```diff
@@ -4,19 +4,16 @@
     pub email: Option<String>,
     #[serde(rename = "publicKey", skip_serializing_if = "Option::is_none")]
-    pub public_key: Option<Box<crate::models::SearchIndexPublicKey>>,
+    pub public_key: Option<Box<crate::models::search_index_public_key::SearchIndexPublicKey>>,
```
See above about potentially removing `Box`

lkatalin replied:
Bump

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,56 @@
+        .await
+        .unwrap();
+    println!("{:#?}", log_entry);
+}
```
I got an error result on this example:
```
[lily@localhost rekor-rs]$ cargo run --example create_log_entry
   Compiling openapi v0.0.1 (/home/lily/Repos/rekor-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 2.24s
     Running `target/debug/examples/create_log_entry`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ResponseError(ResponseContent { status: 409, content: "{\"code\":409,\"message\":\"An equivalent entry already exists in the transparency log with UUID b378b84b79477794b394025d1593a3a1480954c66d51c64ff75eed313cc9f922\"}\n", entity: Some(Status400(Error { code: Some(409), message: Some("An equivalent entry already exists in the transparency log with UUID b378b84b79477794b394025d1593a3a1480954c66d51c64ff75eed313cc9f922") })) })', examples/create_log_entry.rs:54:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

If this is expected, I would add some error handling here to make this error go away - and longer term in another PR we could make sure it can create a new log entry each time.

lkatalin replied:
Bumping my request for error handling in this case, even if we can't fix the error yet.

lkatalin commented on examples/get_log_info.rs:
```diff
@@ -0,0 +1,24 @@
+    let configuration = Configuration::default();
+    let log_info: LogInfo = tlog_api::get_log_info(&configuration).await.unwrap();
+    println!("{:#?}", log_info);
+}
```
I got this error (I think it ran correctly the first few times, though?):
```
[lily@localhost rekor-rs]$ cargo run --example get_log_info
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/examples/get_log_info`
"rootHash: 59ce3532657a4d267161abf5dbba57fb6a71383198c7702bfa2e4d2b6e6fcd71\nsignedTreeHead: |\n  Rekor\n  2449157\n  Wc41MmV6TSZxYav127pX+2pxODGYx3Ar+i5NK25vzXE=\n  Timestamp: 1653505424656916632\n\n  — rekor.sigstore.dev wNI9ajBEAiB9guM7tvmpKSpEY+unOTEmRFUydRgm02qW0iSkQD/i8wIgXHFG+n6dZEAqxaN6yJjQdiv2utcuiFYv7gBFZK1vv88=\ntreeSize: 2449157\n"
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Serde(Error("expected value", line: 1, column: 1))', examples/get_log_info.rs:22:74
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

jyotsna-penumaka replied:
Yeah I noticed this too.
I am not entirely sure why this is happening :thinking: 
![get_log_info_error](https://user-images.githubusercontent.com/60219242/171264162-9e9f4cdd-8c74-47b6-b13e-a9ca899cd853.png)

lkatalin commented on examples/get_public_key.rs:
```diff
@@ -0,0 +1,23 @@
+async fn main() {
+    let configuration = Configuration::default();
+    let pubkey = pubkey_api::get_public_key(&configuration, None).await;
+    println!("{:#?}", pubkey.unwrap());
```
The formatting on this one could be tweaked to be a bit more readable, but not a big deal.

lkatalin replied:
This is fine for this PR but I'd actually suggest opening an issue to make it a bit more readable so we don't forget it!

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,34 @@
+    };
+    let query = SearchIndex {
+        email: Some("jpenumak@redhat.com".to_string()),
+        public_key: Some(Box::new(public_key)),
```
Do you need this `Box`? I tried removing it here and in the model, and the example still worked for me.

jyotsna-penumaka replied:
Yeah you are right :100: 
I don't think we need it :smile:

lkatalin commented on src/models/inclusion_proof.rs:
```diff
@@ -28,7 +25,12 @@ pub struct InclusionProof {
+    pub fn new(
+        log_index: i32,
+        root_hash: String,
+        tree_size: i32,
```
Same as above

lkatalin commented on src/models/inclusion_proof.rs:
```diff
@@ -28,7 +25,12 @@ pub struct InclusionProof {
 impl InclusionProof {
-    pub fn new(log_index: i32, root_hash: String, tree_size: i32, hashes: Vec<String>) -> InclusionProof {
+    pub fn new(
+        log_index: i32,
```
I see this one as `i64` too
https://github.com/sigstore/rekor/blob/main/pkg/sharding/log_index.go#L28

lkatalin reviewed (changes requested):
Overall, this looks amazing and is a lot of great work! :100: It is largely clean and readable and can be matched to the Rekor functionality, and most of the examples worked smoothly for me.

I did get two errors on `create_log_entry` and `get_log_info` - if these are expected, it would be good to handle them and have some output like "This entry already exists" and not show the panic. I'm not sure why I got the other error.

There were some leftover comments in the code that probably need to be removed (or maybe turned into TODO comments, that is more personal preference but I find that easier to come back to). I'm not sure if they referred to stuff you have already done or changes you are planning.

My only other general suggestion for the future would be to make small, incremental commits as much as possible - for example, you could separate out one commit for adding GitHub Actions, one commit for all the `cargo fmt` stuff that is not related to your other changes, etc. just so each commit is a logical whole and easy to read.

A few other small things inline!

flavio commented on examples/get_log_entry_by_uuid.rs:
```diff
@@ -0,0 +1,27 @@
+#[tokio::main]
+async fn main() {
+    let configuration = Configuration::default();
+    let uuid = "073970a07c978b7a9ff15b69fe15d87dfb58fd5756086e3d1fb671c2d0bd95c0";
```
Going back to what I suggested above, I think it would be great to have this value provided by the user via a cli flag. This value could be left as a comment inside of the source code if you want.

flavio commented on examples.sh:
```diff
@@ -0,0 +1,5 @@
+for example in ./examples/*.rs
+do 
+    FILE="$(basename "$example")"
+    cargo run --example "${FILE%.*}" -- $args
```
nit, what is the purpose of this bash script? Given all the examples have hard coded values, I feel like you're treating the examples like end-to-end tests, which IMHO is not their goal 🤔

flavio commented on src/apis/timestamp_api.rs:
```diff
@@ -4,16 +4,22 @@
+use std::os::unix::fs::PermissionsExt;
+use std::path::PathBuf;
 
+const FILENAME: &str = "response.tsr";
```
Do we really need to add this binary blob to the repository? Can't we let the use provide it or have the code fetch it somehow based on user input?

If the binary file must be part of the repo, at least move it to a dedicated folder (dunno `/assets`) instead of having it inside of the root of the repository

jyotsna-penumaka replied:
The timestamp response is saved in the file response.tsr when the user makes `get_timestamp_response` api call.
The user is supposed to provide timestamp request file (an example for the input file is in https://github.com/jyotsna-penumaka/rekor-rs/blob/rekor-functionality/test_data/test_request.tsq )
This file doesn't need to be part of the repo, it is generated each time the user calls `get_timestamp_response` api call.
I can move the place where it is stored to be inside test_data.

flavio commented on src/apis/entries_api.rs:
```diff
@@ -74,25 +84,40 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        // Format the returned response such that it can be read into a struct
+        let uuid: &str = &local_var_content[1..67];
+        let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+        let sum = "{\"uuid\": ".to_string() + uuid + "," + rest;
```
I don't like this way of building the json object. It feels pretty fragile. Can you give me an example of the contents of `local_var_content`?

lkatalin replied:
I had suggested this to Jyotsna initially as a quick but hacky workaround for the [dynamic top-level key problem in Rekor](https://github.com/sigstore/rekor/issues/808), which should be fixed soon.

jyotsna-penumaka replied:
@flavio 
We will be getting rid of parsing all together once https://github.com/sigstore/rekor/issues/808 gets resolved.

We need to modify the response before we can read it into the LogEntry Struct
This is the response that is returned by Rekor (`local_var_content`):

    {
        "9bf4f37447a48848c1f69b6463190cd6cef7728d6b86b7723f08aad0c54cb8d4": {
            "body": "eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoicmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiI0NmJkMzE5ZDM1OTBkMmI0ZDdjN2EyN2M5OWQzMmY3ZWE2MGE4NTBlYzM4MDYzNTFlMDRkMTYxZDAxNGVjYzAxIn19LCJzaWduYXR1cmUiOnsiY29udGVudCI6IkxTMHRMUzFDUlVkSlRpQlRVMGdnVTBsSFRrRlVWVkpGTFMwdExTMEtWVEZPU1ZVd2JFaEJRVUZCUVZGQlFVRkVUVUZCUVVGTVl6Tk9iMHhYVm10TmFsVXhUVlJyUVVGQlFXY3ZkbVZUWXpSdmJIQkxkRTF2VDFJM2NuZG1PQXBXUjBoNmFHaG5NRVpKYjBSMFl6VlNNa3BzZEhwSFowRkJRVUZGV20xc2MxcFJRVUZCUVVGQlFVRkJSMk15YUdoT1ZFVjVRVUZCUVZWM1FVRkJRWFI2Q21NeVozUmFWMUY1VGxSVmVFOVJRVUZCUlVKalEydDBaME14V1dwcmIzZEtkSEJzZVhCRFNEUTJhbkV5UW1Sb05tUjZhblIwZVd0SFpWRjVLMG8xZUhvS2RIRlVlbXRqUXpCWVNHRkZZWEZPZFdNMWNURnpUbEZNWTJRNFNEUjRNM0ZLU2xSRFFsRnZUd290TFMwdExVVk9SQ0JUVTBnZ1UwbEhUa0ZVVlZKRkxTMHRMUzBLIiwiZm9ybWF0Ijoic3NoIiwicHVibGljS2V5Ijp7ImNvbnRlbnQiOiJjM05vTFdWa01qVTFNVGtnUVVGQlFVTXpUbnBoUXpGc1drUkpNVTVVUlRWQlFVRkJTVkEzTTJ0dVQwdEtZVk55VkV0RWEyVTJPRWd2UmxKb09EUlpXVTVDVTB0Qk4xaFBWV1JwV21KamVHOEsifX19fQ==",
            "integratedTime": 1643749079,
            "logID": "c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d",
            "logIndex": 1236200,
            "verification": {
                "signedEntryTimestamp": "MEQCID3QHHXwGauKUfFvs0YCMKZ4e3BE1ZIhrJ5RHnoHGophAiA/R5cDc8JnSUq6F7U/8oTk8/mn7QCdDXI+NRT7qm78+Q=="
            }
        }
    }

Since there is no member called "uuid" in the json body,
we cannot read it into the LogEntry struct.
So we add "{"uuid": " to the returned response and make it a valid json like so:

    {
    	"uuid": "9bf4f37447a48848c1f69b6463190cd6cef7728d6b86b7723f08aad0c54cb8d4",
    	"body": "eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoicmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiI0NmJkMzE5ZDM1OTBkMmI0ZDdjN2EyN2M5OWQzMmY3ZWE2MGE4NTBlYzM4MDYzNTFlMDRkMTYxZDAxNGVjYzAxIn19LCJzaWduYXR1cmUiOnsiY29udGVudCI6IkxTMHRMUzFDUlVkSlRpQlRVMGdnVTBsSFRrRlVWVkpGTFMwdExTMEtWVEZPU1ZVd2JFaEJRVUZCUVZGQlFVRkVUVUZCUVVGTVl6Tk9iMHhYVm10TmFsVXhUVlJyUVVGQlFXY3ZkbVZUWXpSdmJIQkxkRTF2VDFJM2NuZG1PQXBXUjBoNmFHaG5NRVpKYjBSMFl6VlNNa3BzZEhwSFowRkJRVUZGV20xc2MxcFJRVUZCUVVGQlFVRkJSMk15YUdoT1ZFVjVRVUZCUVZWM1FVRkJRWFI2Q21NeVozUmFWMUY1VGxSVmVFOVJRVUZCUlVKalEydDBaME14V1dwcmIzZEtkSEJzZVhCRFNEUTJhbkV5UW1Sb05tUjZhblIwZVd0SFpWRjVLMG8xZUhvS2RIRlVlbXRqUXpCWVNHRkZZWEZPZFdNMWNURnpUbEZNWTJRNFNEUjRNM0ZLU2xSRFFsRnZUd290TFMwdExVVk9SQ0JUVTBnZ1UwbEhUa0ZVVlZKRkxTMHRMUzBLIiwiZm9ybWF0Ijoic3NoIiwicHVibGljS2V5Ijp7ImNvbnRlbnQiOiJjM05vTFdWa01qVTFNVGtnUVVGQlFVTXpUbnBoUXpGc1drUkpNVTVVUlRWQlFVRkJTVkEzTTJ0dVQwdEtZVk55VkV0RWEyVTJPRWd2UmxKb09EUlpXVTVDVTB0Qk4xaFBWV1JwV21KamVHOEsifX19fQ==",
    	"integratedTime": 1643749079,
    	"logID": "c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d",
    	"logIndex": 1236200,
    	"verification": {
    		"signedEntryTimestamp": "MEQCID3QHHXwGauKUfFvs0YCMKZ4e3BE1ZIhrJ5RHnoHGophAiA/R5cDc8JnSUq6F7U/8oTk8/mn7QCdDXI+NRT7qm78+Q=="
    	}
    }

flavio commented on src/apis/entries_api.rs:
```diff
@@ -74,25 +84,40 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
+    configuration: &configuration::Configuration,
+    log_index: i32,
+) -> Result<LogEntry, Error<GetLogEntryByIndexError>> {
     let local_var_configuration = configuration;
```
Why do you need to do that? Both variables will be pointing to the same area or memory 🤔

flavio commented on src/apis/entries_api.rs:
```diff
@@ -102,25 +127,43 @@ pub async fn get_log_entry_by_index(configuration: &configuration::Configuration
+    configuration: &configuration::Configuration,
+    entry_uuid: &str,
+) -> Result<LogEntry, Error<GetLogEntryByUuidError>> {
     let local_var_configuration = configuration;
```
Same here

flavio commented on src/apis/entries_api.rs:
```diff
@@ -52,18 +51,29 @@ pub enum SearchLogQueryError {
+    configuration: &configuration::Configuration,
+    proposed_entry: crate::models::ProposedEntry,
+) -> Result<LogEntry, Error<CreateLogEntryError>> {
     let local_var_configuration = configuration;
```
Why do you need to do that? Both variables are pointing to the same area of memory 🤔

jyotsna-penumaka replied:
I got rid of it now.
Thanks for pointing it out :100:

flavio commented on src/apis/entries_api.rs:
```diff
@@ -130,24 +173,42 @@ pub async fn get_log_entry_by_uuid(configuration: &configuration::Configuration,
+    configuration: &configuration::Configuration,
+    entry: crate::models::SearchLogQuery,
+) -> Result<std::string::String, Error<SearchLogQueryError>> {
     let local_var_configuration = configuration;
```
Same here

flavio commented on src/apis/mod.rs:
```diff
@@ -16,7 +16,7 @@ pub enum Error<T> {
 }
 
-impl <T> fmt::Display for Error<T> {
+impl<T> fmt::Display for Error<T> {
```
I would suggest to remove all this error handling code and use something like the `thiserror` crate to automatically handle everything.

You can see that in action inside of sigstore-rs and oci-distribution

flavio commented on src/models/error.rs:
```diff
@@ -4,13 +4,10 @@
-
-
 #[derive(Clone, Debug, PartialEq, Default, Serialize, Deserialize)]
 pub struct Error {
```
Same tip here, use something like `thiserror` or `snafu`

flavio commented on src/models/search_index_public_key.rs:
```diff
@@ -31,7 +28,7 @@ impl SearchIndexPublicKey {
 }
 
-/// 
+///
```
Missing docs

jyotsna-penumaka commented:
I am not entirely sure why get_info_seems to fail sometimes and pass sometimes.
I made the tree_size field optional, since that field seems to be missing in some logs.
I am attaching a screen shot that shows the uncertainty of running the test.
![get_log_info_error](https://user-images.githubusercontent.com/60219242/171256483-5cdfa84c-3d93-4fc2-8390-96b5103c14a4.png)

jyotsna-penumaka committed:
Remove git_push.sh

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add file to test create_log_entry

Signed-off-by: jyotsna-penumaka <jpenumak@redhat.com>

jyotsna-penumaka committed:
Add cli support to all the examples.

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Run Cargo fmt

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Remove examples running script

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Fix clippy errors

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
- Use thierror crate
- Add documentation for the crate
- Add Readme

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Cargo fmt + clean the code in search_log_query.rs

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
cargo fmt + debug github action

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Update .gitignore, so that it skips response.tsr

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Debug github actions: Test suit

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Use thiserror crate

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Update the data requiered to create a log in reko

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Debug create_log_entry

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

lkatalin commented on README.md:
```diff
@@ -0,0 +1,29 @@
+ ----------------------|------|---------
+ [![Continuous integration](https://github.com/sigstore/sigstore-rs/actions/workflows/tests.yml/badge.svg)](https://github.com/sigstore/sigstore-rs/actions/workflows/tests.yml) | [![Docs](https://img.shields.io/badge/docs-%20-blue)](https://sigstore.github.io/sigstore-rs/sigstore) |  [![License: Apache 2.0](https://img.shields.io/badge/License-Apache2.0-brightgreen.svg)](https://opensource.org/li ...
+ 
+# Rust API client for openapi
```
Is this supposed to be "for Rekor"?

jyotsna-penumaka replied:
Thanks for pointing it out :smile: 
I changed the link for integration tests to point to rekor-rs's integration tests.
The link for the docs is a place holder for now. Once I publish the crate I will have a link to the docs and I will replace it.
https://github.com/sigstore/sigstore-rs/actions/workflows/tests.yml/badge.svg is a graphic that is in sigstore-rs that I also want to use.

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,125 @@
+    create_log_entry will panick if the example code is run,
+    with the following error message:
+
+    thread 'main' panicked at 'called `Result::unwrap()` on an `Err`
+    value: ResponseError(ResponseContent { status: 409, content: "{\"code\":409,\"message\":\
+    "An equivalent entry already exists in the transparency log with UUID
+    1377da9d9dbad451a5a8acdd28add750815d34e8205f1b8a35a67b8a27dae9bf\"}\n",
+    entity: Some(Status400(Error { code: Some(409), message: Some("An equivalent entry
+    already exists in the transparency log with
+    UUID 1377da9d9dbad451a5a8acdd28add750815d34e8205f1b8a35a67b8a27dae9bf") })) })',
+    examples/create_log_entry.rs:104:10
+    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I like this explanation so that users know what to expect with the example code. However, it seems better if instead of letting the program panic, it instead emits a handled error, which is then explained here.

jyotsna-penumaka replied:
I got rid of the unwrap(), so that the code doesn't panic when the user tries to create another entry with the same set of values.

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,125 @@
+    .arg(Arg::new("public_key")
+             .long("public_key")
+             .takes_value(true)
+             .help("base64 encoded public_key. Look at https://raw.githubusercontent.com/jyotsna-penumaka/rekor-rs/rekor-functionality/test_data/create_log_entry.md for more details on generating keys."))
```
:+1:

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,125 @@
+        AlgorithmKind::sha256,
+        flags
+            .value_of("hash")
+            .unwrap_or("43b8f7d99e183c9be264717ddde94e29cfdd5d4b0e8c3906d815f44be54b76f6")
```
Can we make these default values `consts` instead? That way they are easier to handle and update (ex. in the case of the API version)

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,125 @@
+    );
+    let spec = Spec::new(signature, data);
+    let proposed_entry = ProposedEntry::Hashedrekord {
+        api_version: flags.value_of("api_version").unwrap_or("0.0.1").to_string(),
```
Does the user input for API version get checked somewhere to make sure it's supported / not nonsense?

lkatalin replied:
Actually do the user-provided values in general go through some check to make sure they are not malformed, etc.?

jyotsna-penumaka replied:
There is a check on the user provided values on the server side, you get error messages like the following:

`"Error processing entry: invalid value for hash"`
`"Error processing entry: invalid public key: failure decoding PEM"`
`"Error processing entry: verifying signature: failed to verify signature: %!w(<nil>)"`
`"Error processing entry: hashedrekord implementation for version '0.1.1' not found: unable to locate entry for version 0.1.1"`

lkatalin commented on examples/create_log_entry.rs:
```diff
@@ -0,0 +1,125 @@
+
+    let log_entry: LogEntry = entries_api::create_log_entry(&configuration, proposed_entry)
+        .await
+        .unwrap();
```
Does this have to be an unwrap?

lkatalin commented on examples/get_log_info.rs:
```diff
@@ -0,0 +1,36 @@
+    Example command :
+    cargo run --example get_log_info
+
+    The server might return an error sometimes,
+    this is because the result depends on the kind of rekor object that gets returned.
```
This raises question marks for me. How does the golang Rekor handle this?

lkatalin commented on examples/get_timestamp_cert_chain.rs:
```diff
@@ -0,0 +1,35 @@
+timestamping authority has been removed from rekor.
+So this test will fail.
+https://github.com/sigstore/rekor/pull/813
+*/
```
Seems like we should remove this entirely, then? Does this API still exist? (It shouldn't?)

lkatalin commented on examples/get_timestamp_response.rs:
```diff
@@ -0,0 +1,39 @@
+timestamping authority has been removed from rekor.
+So this test will fail.
+https://github.com/sigstore/rekor/pull/813
+*/
```
Remove this example then, please

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -52,18 +51,34 @@ pub enum SearchLogQueryError {
-
-    let local_var_uri_str = format!("{}/api/v1/log/entries", local_var_configuration.base_path);
-    let mut local_var_req_builder = local_var_client.request(reqwest::Method::POST, local_var_uri_str.as_str());
+// Formats the returned response such that it can be read into a struct
+pub fn parse_response(local_var_content: String) -> String {
+    let uuid: &str = &local_var_content[1..67];
+    let rest: &str = &local_var_content[69..local_var_content.len() - 2];
+    "{\"uuid\": ".to_string() + uuid + "," + rest
+}
 
```
Great, this is much better in its own function. I would even still put a TODO comment here with a link out to the Rekor issue so that we can remove this once it's fixed. (And open an issue for it as well.)

lkatalin commented on src/apis/entries_api.rs:
```diff
@@ -74,25 +89,34 @@ pub async fn create_log_entry(configuration: &configuration::Configuration, prop
 
     if !local_var_status.is_client_error() && !local_var_status.is_server_error() {
-        serde_json::from_str(&local_var_content).map_err(Error::from)
+        serde_json::from_str::<LogEntry>(&(parse_response(local_var_content))).map_err(Error::from)
```
So this syntax looked a little bit funny and I checked it the way you had it before and it compiled and ran for me. Is there a reason to make this change?

lkatalin commented on src/apis/mod.rs:
```diff
@@ -1,5 +1,5 @@
-use std::error;
-use std::fmt;
+//Jyotsna's attempt to use the thiserror crate
```
You can remove :)

lkatalin commented on src/apis/timestamp_api.rs:
```diff
@@ -4,16 +4,22 @@
 
-use crate::apis::ResponseContent;
-use super::{Error, configuration};
+// Case 1:
```
Remove please

lkatalin commented on src/apis/timestamp_api.rs:
```diff
@@ -34,60 +40,95 @@ pub enum GetTimestampResponseError {
+        let mut output = fs::File::create(FILENAME)?;
+        let metadata = output.metadata()?;
+        let mut permissions = metadata.permissions();
+        permissions.set_mode(0o644);
```
This mode should maybe also be a const as we might want to update it at some point

lkatalin commented on src/apis/tlog_api.rs:
```diff
@@ -55,30 +56,44 @@ pub async fn get_log_info(configuration: &configuration::Configuration, ) -> Res
-pub async fn get_log_proof(configuration: &configuration::Configuration, last_size: i32, first_size: Option<i32>, tree_id: Option<&str>) -> Result<crate::models::ConsistencyProof, Error<GetLogProofError>> {
+pub async fn get_log_proof(
+    configuration: &configuration::Configuration,
+    last_size: i32,
```
Is there a reason it is i32 and not i64?

lkatalin commented on src/models/log_info.rs:
```diff
@@ -24,21 +21,19 @@ pub struct LogInfo {
 
 impl LogInfo {
-    pub fn new(root_hash: String, tree_size: i32, signed_tree_head: String, tree_id: String) -> LogInfo {
+    pub fn new(root_hash: String, tree_size: i32, signed_tree_head: String) -> LogInfo {
```
i64?

lkatalin commented on src/models/log_info.rs:
```diff
@@ -24,21 +21,19 @@ pub struct LogInfo {
             tree_size,
             signed_tree_head,
-            tree_id,
+            tree_id: None,
```
Why no tree ID?

lkatalin commented on test_data/create_log_entry.md:
```diff
@@ -0,0 +1,33 @@
+How to generate data requiered to upload an artifact on Rekor:
```
I like this reference a lot :+1:

lkatalin commented on examples/get_rekor_version.rs:
```diff
@@ -0,0 +1,30 @@
+    let configuration = Configuration::default();
+    let rekor_version = server_api::get_rekor_version(&configuration).await;
+    println!("{:#?}", rekor_version.unwrap());
+}
```
I've got this error here:

```
[lily@localhost src]$ cargo run --example get_rekor_version
   Compiling rekor v0.0.1 (/home/lily/Repos/rekor-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 2.17s
     Running `/home/lily/Repos/rekor-rs/target/debug/examples/get_rekor_version`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Serde { source: Error("expected value", line: 1, column: 1) }', examples/get_rekor_version.rs:29:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,86 @@
+    --hash e2535d638859bb63ea9ea5cf467562cba63b007eae1acd0d73a3f259c582561f \
+    --public_key c3NoLWVkMjU1MTkgQUFBQUMzTnphQzFsWkRJMU5URTVBQUFBSVA3M2tuT0tKYVNyVEtEa2U2OEgvRlJoODRZWU5CU0tBN1hPVWRpWmJjeG8gdGVzdEByZWtvci5kZXYK \
+    --key_format ssh \
+    --email jpenumak@redhat.com
```
🤔 I'd put in a fake email here... or if you want the user to run this exact command, create a throwaway email

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,86 @@
+    --email jpenumak@redhat.com
+
+
+    The server might return an error sometimes,
+    this is because the result depends on the kind of rekor object that gets returned.
```
What does this mean?

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,86 @@
+
+    let uuid_vec = index_api::search_index(&configuration, query).await;
+    println!("{:#?}", uuid_vec);
+}
```
As noted in the comment, this does return an error for me:

```
[lily@localhost src]$ cargo run --example search_index
   Compiling rekor v0.0.1 (/home/lily/Repos/rekor-rs)
    Finished dev [unoptimized + debuginfo] target(s) in 2.55s
     Running `/home/lily/Repos/rekor-rs/target/debug/examples/search_index`
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', examples/search_index.rs:63:57
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

lkatalin commented on examples/get_log_entry_by_index.rs:
```diff
@@ -0,0 +1,47 @@
+    );
+
+    let flags = matches.get_matches();
+    let index = <i32 as FromStr>::from_str(flags.value_of("log_index").unwrap_or("1")).unwrap();
```
```suggestion
    let index = i32::from_str(flags.value_of("log_index").unwrap_or("1")).unwrap();
```

lkatalin commented on examples/get_log_proof.rs:
```diff
@@ -0,0 +1,59 @@
+
+    let log_proof: ConsistencyProof = tlog_api::get_log_proof(
+        &configuration,
+        <i32 as FromStr>::from_str(flags.value_of("last_size").unwrap_or("10")).unwrap(),
```
```suggestion
        i32::from_str(flags.value_of("last_size").unwrap_or("10")).unwrap(),
```

lkatalin commented on examples/search_index.rs:
```diff
@@ -0,0 +1,86 @@
+
+    let query = SearchIndex {
+        email: Some(flags.value_of("email").unwrap().to_string()),
+        public_key: Some(Box::new(public_key)),
```
```suggestion
        public_key: Some(public_key),
```
You can make this just `Some(public_key)` if you also change the type in `src/models/search_index.rs`

lkatalin commented on examples/search_log_query.rs:
```diff
@@ -0,0 +1,127 @@
+            .value_of("entry_uuids")
+            .unwrap_or("1377da9d9dbad451a5a8acdd28add750815d34e8205f1b8a35a67b8a27dae9bf")
+            .to_string()]),
+        log_indexes: Some(vec![<i32 as FromStr>::from_str(
```
```suggestion
        log_indexes: Some(vec![i32::from_str(
```

lkatalin commented on src/models/search_index.rs:
```diff
@@ -4,19 +4,16 @@
     pub email: Option<String>,
     #[serde(rename = "publicKey", skip_serializing_if = "Option::is_none")]
-    pub public_key: Option<Box<crate::models::SearchIndexPublicKey>>,
+    pub public_key: Option<Box<crate::models::search_index_public_key::SearchIndexPublicKey>>,
```
```suggestion
    pub public_key: Option<crate::models::search_index_public_key::SearchIndexPublicKey>,
```

lkatalin reviewed (changes requested):
I like that you provided default values for the examples but also let the user input their own! I used the default so I'm glad you left these. : )

Comments inline, but in general I would like changes on:
- Error handling (instead of letting the code panic)
- Removing `unwrap()`s (should help with above)
- Sanity-checking the user inputs for the commands
- Really checking the `i32` vs `i64` values because I see both
- Moving the default values to `const`s
- Not letting the requests just return an error... makes sense for `create_entry` and we can fix that later, but the other ones I don't understand

Not sure if this summary covers everything but you will see the comments inline.

jyotsna-penumaka committed:
Move the default values to consts

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Run cargo fmt

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Remove redundant code

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>

jyotsna-penumaka committed:
Remove timestamping authority API and examples.

timestamping authority has been removed from rekor.
So this test will fail.
https://github.com/sigstore/rekor/pull/813

Signed-off-by: jyotsna-penumaka <jyotsnap@bu.edu>
